examples/pytorch/graphsage/train_sampling.py

```diff
@@ -52,13 +52,17 @@ def run(args, device, data):
     val_nid = th.nonzero(val_g.ndata['val_mask'], as_tuple=True)[0]
     test_nid = th.nonzero(~(test_g.ndata['train_mask'] | test_g.ndata['val_mask']), as_tuple=True)[0]
 
-    dataloader_device = th.device('cpu')
-    if args.sample_gpu:
+    if args.sample_device == 'gpu':
         train_nid = train_nid.to(device)
         # copy only the csc to the GPU
         train_g = train_g.formats(['csc'])
         train_g = train_g.to(device)
-        dataloader_device = device
+        args.num_workers = 0
+    elif args.sample_device == 'uva':
+        train_nid = train_nid.to(device)
```

Comment:
Would this impact performance significantly if `train_nid` were a unified tensor rather than being stored on the device?

Reply:
@nv-dlasalle In fact `train_nid` has to be on the GPU because we use the context of it (or `seed_nodes` more precisely) to execute the sampling kernel.

Reply:
Actually, I feel that the device switching for sampling and subgraph construction (`to_block`) is a bit complicated now. I summarize it as a table and plan to put it in the doc.

<table>
<thead>
  <tr>
    <th>graph device</th>
    <th>nids device</th>
    <th>dataloader device</th>
    <th>num_workers</th>
    <th>sampling</th>
    <th>subgraph construction</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td colspan="2">GPU</td>
    <td>GPU</td>
    <td>0</td>
    <td>GPU</td>
    <td>GPU</td>
  </tr>
  <tr>
    <td colspan="2" rowspan="2">CPU</td>
    <td rowspan="2">GPU</td>
    <td>0</td>
    <td>CPU</td>
    <td>GPU</td>
  </tr>
  <tr>
    <td>&gt;0</td>
    <td rowspan="2">CPU</td>
    <td>CPU</td>
  </tr>
  <tr>
    <td colspan="2">CPU</td>
    <td>CPU</td>
    <td>any</td>
    <td>CPU</td>
  </tr>
  <tr>
    <td>Pinned</td>
    <td>GPU</td>
    <td>GPU</td>
    <td>0</td>
    <td>GPU</td>
    <td>GPU</td>
  </tr>
</tbody>
</table>

Reply:
While we may have to do some tuning to ensure its faster in all cases, I'd be in favor of deprecating the normal sample-on-cpu to-block-on-gpu in favor UVA sampling, to keep things simple.

Reply:
Should we include it in this PR? Another question is what to do with the `output_device` argument in `BlockSampler`? I think it is introduced to enable the hybrid mode.

Reply:
I think we could simplify it as:

<html><body>
<!--StartFragment-->

graph device | nids device | dataloader device | num_workers | sampling | subgraph construction
-- | -- | -- | -- | -- | --
GPU | GPU | GPU | 0 | GPU | GPU
CPU | CPU | CPU | any | CPU | CPU
Pinned | GPU | GPU | 0 | GPU | GPU

<!--EndFragment-->
</body>
</html>

We could also simplify this further, by having the DataLoader copy the nids to the proper device for each mini-batch, especially since relative to the output of the mini-batch they should be quite small.